File: internal/services/create_user_service.go
package services

import (
    "errors"

    db "github.com/LuisHenriqueFA14/WizBuy/internal/database"
    "github.com/LuisHenriqueFA14/WizBuy/internal/models"
    "github.com/LuisHenriqueFA14/WizBuy/internal/utils"

    "github.com/google/uuid"
    "golang.org/x/crypto/bcrypt"
)

func CreateUserService(name, email, password string) ([]byte, error) {
    // Email verification
    emailAlreadyInUse := db.DB.Where("email = ?", email).First(&models.User{})

    if emailAlreadyInUse.RowsAffected > 0 {
        return nil, errors.New("Email already in use")
    }

    if !utils.IsValidEmail(email) {
        return nil, errors.New("Invalid email")
    }

    // Password verification
    if !utils.IsValidPassword(password) {
        return nil, errors.New("Invalid password")
    }

    hashedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)

    if err != nil {
        return nil, errors.New("Error encrypting password")
    }

    user := models.User{
        Id: uuid.NewString(),
        Name: name,
        Email: email,
        Password: string(hashedPassword),
    }

    db.DB.Create(&user)

    return []byte(user.Id), nil
}

File: src/services/createUserService.js
const { database } = require('../database');
const bcrypt = require('bcrypt');

const { isValidEmail, isValidPassword } = require('../utils/validate');

async function createUserService({ name, email, password }, db = database) {
    if(!name || !email || !password) {
        return {
            error: {
                type: 'BadRequest',
                message: 'Name, email and password are required'
            }
        };
    }

    if(!isValidEmail(email)) {
        return {
            error: {
                type: 'BadRequest',
                message: 'Invalid email'
            }
        };
    }

    if(!isValidPassword(password)) {
        return {
            error: {
                type: 'BadRequest',
                message: 'Invalid password'
            }
        };
    }

    const emailInUse = await db.user.findFirst({
        where: {
            email
        }
    });

    if(emailInUse) {
        return {
            error: {
                type: 'BadRequest',
                message: 'Email already in use'
            },
        };
    }

    const hashedPassword = await bcrypt.hash(password, 8);

    if(!hashedPassword) {
        return {
            error: {
                type: 'InternalServerError',
                message: 'Failed to hash password'
            }
        };
    }

    const user = await db.user.create({
        data: {
            name,
            email,
            password: hashedPassword
        }
    });

    if(!user) {
        return {
            error: {
                type: 'InternalServerError',
                message: 'Failed to create user'
            }
        };
    }

    return {
        response: {
            message: 'User created successfully',
            id: user.id
        }
    };
}

module.exports = { createUserService };
